[LFI] Introduce AArch64 LFI Target

zyedidia opened the pull request:
This PR is the first step towards introducing LFI into LLVM as a new sub-architecture backend of AArch64. For details, please see the [RFC](https://discourse.llvm.org/t/rfc-lightweight-fault-isolation-lfi-efficient-native-code-sandboxing-upstream-lfi-target-and-compiler-changes/88380), which has been approved for AArch64.

This patch creates the `aarch64_lfi` architecture, and marks the appropriate registers as reserved when it is targeted (`x25`, `x26`, `x27`, `x28`). It also adds a Clang driver toolchain for targeting LFI, and updates the compiler-rt CMake to allow builds for the `aarch64_lfi` target. The patch also includes documentation for LFI and the rewrites that will be implemented in future patches.

I am planning to split the relevant modifications for LFI into a series of patches, organized as described below (after this one). Please let me know if you'd like me to split the changes in a different way, or provide one big patch.

1. The next patch will introduce the `MCLFIExpander` mechanism for applying the MC-level rewrites needed by LFI, along with the `.lfi_expand` and `.lfi_no_expand` assembly directives when targeting LFI. A preview can be seen on the `lfi-project` [fork](https://github.com/llvm/llvm-project/compare/main...lfi-project:llvm-project:lfi-patchset/aarch64-pr-2).

2. The following patch will create an `MCLFIExpander` for the AArch64 backend that performs LFI expansions. This patch will contain the majority of the LFI-specific logic.

3. The final patch will add an optimization to the rewriter that can eliminate redundant guard instructions that occur within the same basic block.

We plan to introduce x86-64 support after further discussion and once the `MCLFIExpander` infrastructure is in place.

Please let me know your feedback, and thank you very much for your help and guidance in the review process.

zyedidia committed:
[LFI] Add aarch64_lfi subtarget

zyedidia committed:
[LFI] Reserve LFI registers

zyedidia committed:
[LFI] Add Clang driver toolchain for LFI

zyedidia committed:
[LFI] Update compiler-rt cmake for LFI

zyedidia committed:
[LFI] Add LFI.rst documentation

zyedidia committed:
[LFI] Apply clang-format to Driver.cpp

zyedidia committed:
[LFI] Fix title underline in docs/LFI.rst

zyedidia committed:
[LFI] Mention LFI backend in CodeGenerator.rst

zyedidia committed:
[LFI] Add LFI to UserGuides toctree

zyedidia committed:
Merge branch 'main' into lfi-patchset/aarch64-pr-1

davemgreen commented on llvm/docs/CodeGenerator.rst:
```diff
@@ -2492,3 +2492,10 @@ The AMDGPU backend
 directory. This code generator is capable of targeting a variety of
 AMD GPU processors. Refer to :doc:`AMDGPUUsage` for more information.
+
+The Lightweight Fault Isolation (LFI) backend
```
There is not expected to be a LFI backend, right? It is a modification to the existing backends that support it.

davemgreen commented on llvm/docs/LFI.rst:
```diff
@@ -0,0 +1,387 @@
+* Performance: LFI aims for minimal overhead vs. unsandboxed code.
+* Security: The LFI runtime and compiler elements aim to be simple and
+  verifiable when possible.
+* Usability: LFI aims to make it easy as possible to used retrofit sandboxing,
```
to use? or to be used for?

smithp35 commented on llvm/docs/LFI.rst:
```diff
@@ -0,0 +1,387 @@
+technique of Software-Based Fault Isolation (SFI). Initial support for LFI in
+LLVM is focused on the AArch64 platform, with x86-64 support planned for the
+future. The initial version of LFI for AArch64 is designed to support the
+Armv8.1 AArch64 architecture.
```
I think it would be useful to set out what the minimum and maximum are here. If I've understood LFI correctly then the maximum is determined by what the verifier can understand, the minimum could in theory be v8-A but the platform may insist on a minium of v8.1-m, presumably to get inline LSE atomics.

What is expected to happen if someone does  `--target=aarch64_lfi-linux-musl` with an out of range target or set of features? Does it give an error message or does it silently remove incompatible features?

smithp35 commented on llvm/docs/LFI.rst:
```diff
@@ -0,0 +1,387 @@
+
+These rewrites (also called "expansions") are applied at the very end of the
+LLVM compilation pipeline (during the assembler step). This allows the rewrites
+to be applied to hand-written assembly, including inline assembly.
```
Although likely to be rare. I expect that there are some assembly examples that cannot be rewritten. I assume in that case, the user gets a "incompatible with LFI, please rewite your program" error message.

zyedidia replied:
Yes, these are mostly cases of instructions that use reserved registers, or that we don't have safe rewrites for (scatter-gather or privileged instructions for example). I think down the line we can introduce mechanisms to rewrite uses of reserved registers to spill to a thread-local memory location though.

smithp35 reviewed (commented):
My apologies for the delay, managed to catch me at a busy time in the day job. Will be out of office till next week so may be a bit slow to reply.

I appreciate that you've included the documentation in the patch, I've put some questions there that probably don't affect this patch, but may affect later ones. I can't spot anything obviously wrong in this patch.

The one important decision in this patch is the choice of a subtarget for LFI. I guess the most obvious alternatives would be a LFI envionment (like pauthtest), or an ABI (like aapcs-soft). LFI is a bit strange in that it is kind of a mixture of all three. The instructions are restricted to what the verifier can accept, the syscalls are limited to what the sandbox accepts, and there is a different procedure call standard. Given that objects compiled for LFI won't be compatible with non-LFI I think a sub-target is probably the best choice, or at least I can't think of a better one.

Your suggestions of how to split the patch sound reasonable to me. Without spending a lot of effort looking at the full patch, it is diffcult to know if there is a significantly better way. I guess we'll let you know if it needs splitting again.

I expect to have more comments on later patches. Particularly if there's anything we need to document in the Arm ABI. For example identifying relocatable objects as LFI (to diagnose clashes with non-LFI at link-time) and identifying executables as requiring LFI.

zyedidia commented on llvm/docs/LFI.rst:
```diff
@@ -0,0 +1,387 @@
+|    ldur xN, [x28, #16]    |                           |
+|                           |                           |
++---------------------------+---------------------------+
+
```
That's a nice point, thanks! We can have an optimization that directly places the result of `adrp` into `x28` so that it can be used without a further guard with `ldr`:

```
adrp x28, target
ldr xN, [x28, imm]
```

I made an initial implementation of this yesterday that appears to provide a small (~0.5 percentage point) performance improvement.

zyedidia commented on llvm/docs/LFI.rst:
```diff
@@ -0,0 +1,387 @@
+
+The LFI target has several configuration options.
+
+* ``+lfi-stores``: create a "stores-only" sandbox, where rewrites are not applied to loads.
```
Ok that seems reasonable to me, I agree these can get a bit confusing since they stack. How about these as the features:

* `+nolfi-loads`: disables sandboxing for loads.
* `+nolfi-stores`: disables sandboxing for stores.

So if you want the `jumps-only` mode you would use `+nolfi-loads+nolfi-stores` and `stores-only` would be `+nolfi-loads`.

smithp35 replied:
Thanks for the suggestion, that sounds good to me.

zyedidia committed:
[LFI] Adjust LFI toolchain definition

Sets libc++ as the default libcxx for LFI, and uses superclass for
setting libcxx arguments.

smithp35 reviewed (commented):
Thanks for the updates. I've resolved most of the conversations.

One other thing that could be worth considering for the future is how the rewrites interact with linker optimization (relaxation) https://github.com/ARM-software/abi-aa/blob/main/aaelf64/aaelf64.rst#relocation-optimization 

To some extent this is just delaying some rewriting until link time, it would only make sense if the extra information available at link time, such as the destination address of targets would permit better code-sequences.

zyedidia committed:
[LFI] Update LFI.rst documentation

Adds info about the adrp optimization and updates plan for
lfi-stores/lfi-loads features.

zyedidia committed:
[LFI] Update docs based on feedback

zyedidia commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -6864,6 +6865,9 @@ const ToolChain &Driver::getToolChain(const ArgList &Args,
       else if (Target.isWALI())
         TC = std::make_unique<toolchains::WebAssembly>(*this, Target, Args);
+      else if (Target.isLFI())
+        TC = std::make_unique<toolchains::LFILinuxToolChain>(*this, Target,
```
That sounds fine to me. I wasn't sure what to go with in this case because there are other toolchains with `ToolChain` in the same (`toolchains::PPCLinuxToolChain`, `toolchains::HexagonToolChain`, etc.), but I can go ahead and change it.

AaronBallman reviewed (approved):
Aside from the naming nit, the Clang bits LGTM, but please wait for compiler-rt and llvm folks to sign off as well.

zyedidia committed:
[LFI] Rename LFILinuxToolChain to LFILinux

smithp35 reviewed (approved):
Thanks for the updates.

I'm happy with what's in this particular patch. I think the documentation for target support can be updated later as more architecture support is available in the verifier.

zyedidia commented:
Great, thanks! Let me know if we need to request any more reviewers. I will open the next PR once this one is merged.
